blogs/qserial.md

```diff
@@ -0,0 +1,243 @@
+    var isPromiseLike = function (obj) { return obj && angular.isFunction(obj.then); }
+
+    /*
+     * @description Execute a collection of tasks serially.  A task is a function that returns a promise
+     *
+     * @param {Array.<Function>|Object.<Function>} tasks An array or hash of tasks.  A tasks is a function
+     *   that returns a promise.  You can also provide a collection of objects with a success tasks, failure task, and/or notify function
+     * @returns {Promise} Returns a single promise that will be resolved or rejected when the last task
+     *   has been resolved or rejected.
+     */
+    function serial(tasks) {
+      //Fake a "previous task" for our initial iteration
+      var prevPromise;
+      var error = new Error();
+      angular.forEach(tasks, function (task, key) {
+        var success = task.success || task;
+        var fail = task.fail;
+        var notify = task.notify;
+        var nextPromise;
+
+        //First task
+        if (!prevPromise) {
+          nextPromise = success();
+          if (!isPromiseLike(nextPromise)) {
+            error.message = "Task " + key + " did not return a promise.";
+            throw error;
+          }
+        } else {
+          //Wait until the previous promise has resolved or rejected to execute the next task
+          nextPromise = prevPromise.then(
+            /*success*/function (data) {
+              if (!success) { return data; }
+              var ret = success(data);
+              if (!isPromiseLike(ret)) {
+                error.message = "Task " + key + " did not return a promise.";
+                throw error;
+              }
+              return ret;
+            },
+            /*failure*/function (reason) {
+              if (!fail) { return $delegate.reject(reason); }
+              var ret = fail(reason);
+              if (!isPromiseLike(ret)) {
+                error.message = "Fail for task " + key + " did not return a promise.";
+                throw error;
+              }
+              return ret;
+            },
+            notify);
+        }
+        prevPromise = nextPromise;
+      });
+
+      return prevPromise;
+    }
+
+    $delegate.serial = serial;
+    return $delegate;
+  });
+});
+</pre>
+<br/>
+Now that we have the full $q.serial API, let's investigate its power.  $q.serial is most useful when you have a tasks that are expensive, tie up resources, and are not time sensitive.
+
+At <a href="http://hurdlr.com/">Hurdlr</a> we encountered this exact criteria the other day.  We build a mobile app.  Internet tends to be less reliable on phones than on computers.  So we try to minimize server calls.  
+
+One way we minimize server calls is by eagerly loading bulk data.  So depending on module, user's config, etc. we execute a set of tasks that load bulk data.  We don't want to execute all the tasks at once because it can tie up more database resources and phone resources when caching than we'd like.
+ 
```

Comment:
This is one.  Another we've encountered is rendering lists on the phone.  We don't want to render lists too aggressively because they can take a lot of resources.  It's fine to take a few extra milliseconds if the list is hidden behind a loading screen.  So we use $q.serial for rendering list rows where rending each list row is a task.